sympy/ntheory/factor_.py

```diff
@@ -252,12 +252,16 @@ def multiplicity(p, n):
 
     >>> from sympy.ntheory import multiplicity
     >>> from sympy.core.numbers import Rational as R
+    >>> from sympy import factorial
     >>> [multiplicity(5, n) for n in [8, 5, 25, 125, 250]]
     [0, 1, 2, 3, 3]
     >>> multiplicity(3, R(1, 9))
     -2
+    >>> multiplicity(5, factorial(60))
```

Comment:
You should increase that factorial argument to `2**60` and say that if you pass an unevaluated factorial that the multiplicity will be determined without computing the factorial
```python
>>> multiplicity(5, factorial(2**60, evaluate=False))
288230376151711731
```

Reply:
then i will have to call multiplicity_in_factorial if i get a factorial object as the argument of multiplicity

Reply:
Because a factorial object(the unevaluated factorial) will raise ValueError in multiplicity,

Reply:
Yes, you have to detect that case, e.g.
```diff
diff --git a/sympy/ntheory/factor_.py b/sympy/ntheory/factor_.py
index db31c51..9a137f0 100644
--- a/sympy/ntheory/factor_.py
+++ b/sympy/ntheory/factor_.py
@@ -278,6 +278,11 @@ def multiplicity(p, n):
                     multiplicity(p.q, n.p),
                     multiplicity(p.p, n.q))
                 return like - cross
+        elif (isinstance(p, SYMPY_INTS) and
+                isinstance(n, factorial) and
+                isinstance(n.args[0], Integer) and
+                n.args[0] >= 0):
+            return multiplicity_in_factorial(p, n.args[0])
         raise ValueError('expecting ints or fractions, got %s and %s' % (p, n))

     if n == 0:
```

Reply:
You will have to import factorial, however.

Reply:
I also updated the use of digits (and tested it) [here](https://github.com/sympy/sympy/pull/18671#discussion_r380262156).

Reply:
> You will have to import factorial, however.

Not able to import, it is giving some strange error.

Reply:
> I also updated the use of `digits` [here](https://github.com/sympy/sympy/pull/18671#discussion_r380262156). I tested it with
> 
so do I have to change this part in multiplicity_in_factorial and def a new function for it or something else?
> ```python
> for do in range(1000):
>   p = randint(2,10000)
>   n = randint(100,10000)
>   assert minf(p,n) == multiplicity(p, factorial(n))==multiplicity_in_factorial(p,n)
> ```

 And I will have to add this regression test as well in test_factor_.py??

Reply:
No. A few tests chosen to test p with different factoring structures should be fine, e.g.
```python
n = factorial(1000)
for i in (2, 4, 6, 12, 30, 36, 48, 60, 72, 96):
  assert multiplicity(i, n) == multiplicity_in_factorial(i, 1000)
```
Here is a diff in which I further optimized the function by considering only the largest factor with a given multiplicity. Does that look right to you?
```diff
diff --git a/sympy/ntheory/__init__.py b/sympy/ntheory/__init__.py
index 658a0c3..c595b46 100644
--- a/sympy/ntheory/__init__.py
+++ b/sympy/ntheory/__init__.py
@@ -6,7 +6,8 @@
     randprime, Sieve, sieve, primorial, cycle_length, composite, compositepi
 from .primetest import isprime
 from .factor_ import divisors, proper_divisors, factorint, multiplicity, \
-    multiplicity_in_factorial, perfect_power, pollard_pm1, pollard_rho, primefactors, totient, trailing, \
+    multiplicity_in_factorial, perfect_power, pollard_pm1, pollard_rho, \
+    primefactors, totient, trailing, \
     divisor_count, proper_divisor_count, divisor_sigma, factorrat, \
     reduced_totient, primenu, primeomega, mersenne_prime_exponent, \
     is_perfect, is_mersenne_prime, is_abundant, is_deficient, is_amicable, \
@@ -31,12 +32,12 @@
     'isprime',
 
 
-    'divisors', 'proper_divisors', 'factorint', 'multiplicity', 'multiplicity_in_factorial', 'perfect_power',
+    'divisors', 'proper_divisors', 'factorint', 'multiplicity', 'perfect_power',
     'pollard_pm1', 'pollard_rho', 'primefactors', 'totient', 'trailing',
     'divisor_count', 'proper_divisor_count', 'divisor_sigma', 'factorrat',
     'reduced_totient', 'primenu', 'primeomega', 'mersenne_prime_exponent',
     'is_perfect', 'is_mersenne_prime', 'is_abundant', 'is_deficient', 'is_amicable',
-    'abundance', 'dra', 'drm',
+    'abundance', 'dra', 'drm', 'multiplicity_in_factorial'
 
     'npartitions',
 
diff --git a/sympy/ntheory/factor_.py b/sympy/ntheory/factor_.py
index d194d37..ead9046 100644
--- a/sympy/ntheory/factor_.py
+++ b/sympy/ntheory/factor_.py
@@ -3,6 +3,7 @@
 """
 from __future__ import print_function, division
 
+from collections import defaultdict
 import random
 import math
 
@@ -252,16 +253,27 @@ def multiplicity(p, n):
 
     >>> from sympy.ntheory import multiplicity
     >>> from sympy.core.numbers import Rational as R
-    >>> from sympy import factorial
     >>> [multiplicity(5, n) for n in [8, 5, 25, 125, 250]]
     [0, 1, 2, 3, 3]
     >>> multiplicity(3, R(1, 9))
     -2
-    >>> multiplicity(5, factorial(60))
-    14
 
-    """
+    Note: when checking for the multiplicity of a number in a
+    large factorial it is most efficient to send it as an unevaluated
+    factorial or to call ``multiplicity_in_factorial`` directly:
 
+    >>> from sympy.ntheory import multiplicity_in_factorial
+    >>> from sympy import factorial
+    >>> p = factorial(25)
+    >>> n = 2**100
+    >>> nfac = factorial(n, evaluate=False)
+    >>> multiplicity(p, nfac)
+    52818775009509558395695966887
+    >>> _ == multiplicity_in_factorial(p, n)
+    True
+
+    """
+    from sympy.functions.combinatorial.factorials import factorial
     try:
         p, n = as_int(p), as_int(n)
     except ValueError:
@@ -282,6 +294,11 @@ def multiplicity(p, n):
                     multiplicity(p.q, n.p),
                     multiplicity(p.p, n.q))
                 return like - cross
+        elif (isinstance(p, (SYMPY_INTS, Integer)) and
+                isinstance(n, factorial) and
+                isinstance(n.args[0], Integer) and
+                n.args[0] >= 0):
+            return multiplicity_in_factorial(p, n.args[0])
         raise ValueError('expecting ints or fractions, got %s and %s' % (p, n))
 
     if n == 0:
@@ -315,44 +332,67 @@ def multiplicity(p, n):
     return m
 
 
-def multiplicity_in_factorial(b, n):
-    """
-    Finds the greatest integer m such that b**m divides factorial of n without calculating the factorial
+def multiplicity_in_factorial(p, n):
+    """return the largest integer ``m`` such that ``p**m`` divides ``n!``
+    without calculating the factorial of ``n``.
 
     Examples
     ========
 
     >>> from sympy.ntheory import multiplicity_in_factorial
+    >>> from sympy import factorial
 
-    >>> multiplicity_in_factorial(5, 60)
-    14
-    >>> multiplicity_in_factorial(5, 25)
+    >>> multiplicity_in_factorial(2, 3)
+    1
+
+    An instructive use of this is to tell how many trailing zeros
+    a given factorial has. For example, there are 6 in 25!:
+
+    >>> factorial(25)
+    15511210043330985984000000
+    >>> multiplicity_in_factorial(10, 25)
     6
-    >>> multiplicity_in_factorial(12, 49)
-    22
-    >>> multiplicity_in_factorial(10, 100)
-    24
-    >>> multiplicity_in_factorial(42, 10**10)
-    1666666660
 
-    """
+    For large factorials, it is much faster/feasible to use
+    this function rather than computing the actual factorial:
 
-    b = as_int(b)
-    n = as_int(n)
+    >>> multiplicity_in_factorial(factorial(25), 2**100)
+    52818775009509558395695966887
 
-    if b <= 0:
-        raise ValueError('expecting positive integer got %s' % b )
+    """
+    p = as_int(p)
+    n = as_int(n)
 
+    if p <= 0:
+        raise ValueError('expecting positive integer got %s' % p )
     if n < 0:
         raise ValueError('expecting non-negative integer got %s' % n )
 
-    factors = factorint(b)
-
-    min_e = min((n - (sum(digits(n, p)) - p))//(p - 1) for p in factors)
-    # the exponent of prime in n! = (n - sum(digits of n in base `prime`)/(prime - 1)
-    # subtracting prime since digits include prime as first element
-
-    return min_e//min(factors.values())
+    factors = factorint(p)
+
+    # keep only the largest of a given multiplicity since those
+    # of a given multiplicity will be goverened by the behavior
+    # of the largest factor
+    test = defaultdict(int)
+    for k, v in factors.items():
+        test[v] = max(k, test[v])
+    keep = set(test.values())
+    # remove others from factors
+    for k in list(factors.keys()):
+        if k not in keep:
+            factors.pop(k)
+
+    mp = S.Infinity
+    for i in factors:
+        # multiplicity of i in n! is
+        mi = (n - (sum(digits(n, i)) - i))//(i - 1)
+        # multiplicity of p in n! depends on multiplicity
+        # of prime `i` in p, so we floor divide by factors[i]
+        # and keep it if smaller than the multiplicity of p
+        # seen so far
+        mp = min(mp, mi//factors[i])
+
+    return mp
 
 
 def perfect_power(n, candidates=None, big=True, factor=True):
```